### DISCUSSION TITLE
adding libxml to hurl_core adds system dependency

### DISCUSSION THREAD

**Discussion by SilenLoc** [Orange-OpenSource/hurl#2025] [General] [answered]
Thank you for trying to address all of the questions etc.
It is important for me that you do not take this as a critique.

Was it an explicit decision to introduce a library that has a system dependency?

https://github.com/Orange-OpenSource/hurl/commit/f82ed05d9a18af23f5e7cc3361cb841762c50aae

Before this commit, I was able to compile for wasm32 without problems, now hurl_core needs a system dependency that obv. does not work because [libxml](https://github.com/KWARC/rust-libxml/blob/master/Cargo.toml) does not support wasm32.

if it was explicit than I also plan with this and will target only, windows, mac and unix from now on in my repo. Otherwise I wait :)

**Comment by jcamiel** [comment, marked as answer]
Hi @SilenLoc 

There's no bad question of course and we're open to discuss everything!

Yes it was an explicit decision, we think that it'll not be an issue, as `libxml` being already a mandatory dependency of the `hurl` binary (used for XPath evaluation). Let me just explained the reason of this change.

With this change, we would like: 

1. to detect the end of XML document in a performant way
2. remove a dependency on [`xsd-document`](https://github.com/shepmaster/sxd-document) Rust crate which has not been updated for 3 years

The first point deals with this hurl file:

```hurl
POST https://example.org/InStock
Content-Type: application/soap+xml; charset=utf-8
SOAPAction: "http://www.w3.org/2003/05/soap-envelope"
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope" xmlns:m="http://example.net">
  <soap:Header></soap:Header>
  <soap:Body>
    <m:GetStockPrice>
      <m:StockName>GOOG</m:StockName>
    </m:GetStockPrice>
  </soap:Body>
</soap:Envelope>
HTTP 200
```

With the XML body being inline without "strong marker" (that allows a simple and clean file), we need to identify the start and the end of the XML body, at the byte level. Not a lot of XML library allows this kind of parsing, `libxml` being one with a SAX parser. Our previous implementation was to check byte after byte if the read document is  valid XLML so it was really not efficient.

The second point was more to prevent being stuck with a dependency that will be not updated. As `libxml` is "pervasive", we felt more confident to use it instead of the [`xsd-document`](https://github.com/shepmaster/sxd-document).

Now, things could change in the future, as we really like to have as little as dependency as possible. We may change our implementation and do the XML parsing ourself. It's not a priority, it will not be done in the next month but there is a possibility we remove the need to have `libxml` as a dependency of `hurl_core`. 

See also the [PR comment with performance test](https://github.com/Orange-OpenSource/hurl/pull/1910)

**Reply by SilenLoc** [reply]
Wow this is a huge performance improvement ^^

Thank you for explaining.

**Comment by fabricereix** [comment]
It's interesting to know that `hurl_core` is used by itself without `hurl`.

**Reply by SilenLoc** [reply]
well, it is for my personal/supportive project and the functionality of hurl_core is already enough to create tooling I miss for hurl (Or to put it better "I dislike the tooling that is around except for hurl itself"). Not sure if this qualifies as "used by itself" :)

I use it to create a small "Text Editor" that I use from time to time to fix my hurl files. I call it Hurlalot. I would have used all of hurl if all of the packages were wasm32 compatible. Which is obviously not possible with a cli application.

But now I am going to create a server that can parse the files and abstract hurl away from the client.
Like that, it also opens up the rest of hurl for my project.

**Reply by jcamiel** [reply]
A kind of playground for Hurl? Very interesting!

**Reply by SilenLoc** [reply]
yes, I saw you starred it ^^, "playground" is the state it has right now. But it should evolve into a feature-rich editor.
